harness/determined/pytorch/experimental/_torch_batch_process.py

```diff
@@ -0,0 +1,447 @@
+            "a batch_sampler automatically."
+        )
+    if batch_size is not None:
+        if "batch_size" in dataloader_kwargs:
+            raise ValueError(
+                "batch_size is passed into torch_batch_process " "and dataloader_kwargs"
+            )
+
+
+def _validate_iterate_length(iterate_length: Optional[int], times_iterate: int) -> int:
+    if iterate_length is None:
+        return times_iterate
+
+    if iterate_length <= 0:
+        warnings.warn(
+            f"iterate_length {iterate_length} is not valid. "
+            f"Ignoring this argument and iterate over entire dataset once",
+            stacklevel=2,
+        )
+        return times_iterate
+
+    if iterate_length > times_iterate:
+        warnings.warn(
+            f"iterate_length {iterate_length} exceeds sharded dataset length. "
+            f"Ignoring this argument and iterate over entire dataset once",
+            stacklevel=2,
+        )
+        return times_iterate
+    return iterate_length
+
+
+def _get_storage_information(
+    checkpoint_config: Dict[str, Any], default_uuid_path: str, core_context: core.Context
+) -> str:
+    storage_type = checkpoint_config["type"]
+
+    if storage_type == "s3":
+        bucket = checkpoint_config["bucket"]
+        return f"s3://{bucket}/{default_uuid_path}"
+    elif storage_type == "gcs":
+        bucket = checkpoint_config["bucket"]
+        return f"gs://{bucket}/{default_uuid_path}"
+    elif storage_type == "azure":
+        container = checkpoint_config["container"]
+        return f"Azure container: {container} Directory:{default_uuid_path}"
+    elif storage_type == "shared_fs":
+        base_path = core_context.checkpoint._storage_manager._base_path
+        return f"{base_path}/{default_uuid_path}"
+    else:
+        raise NotImplementedError(f"Storage type {storage_type} support is not implemented")
+
+
+def _reduce_metrics(
+    batch_processor_context: TorchBatchProcessorContext,
+    core_context: core.Context,
+    rank: int,
+    steps_completed: int,
+) -> None:
+    reducables = list(batch_processor_context._wrapped_reducers)
+    # If the user has set metric reducers
+    if len(reducables) > 0:
+        # Reduce metrics (blocking as reduce across slots is needed)
+        # Report reduced metrics to master
+        gatherables = [wrapped.per_slot_reduce() for wrapped in reducables]
+        if rank == 0:
+            gathered = core_context.distributed.gather(gatherables)
+            if gathered is not None:
+                metrics = batch_processor_context.run_cross_slot_reduction(reducables, gathered)
+                core_context.train.report_validation_metrics(
+                    steps_completed=steps_completed,
+                    metrics=metrics,
+                )
+        else:
+            # Other ranks sent metrics to chief
+            core_context.distributed.gather(gatherables)
+
+
+def torch_batch_process(
+    batch_processor_cls: Type[TorchBatchProcessor],
```

Comment:
I find this part of the interface unexpected. In an ideal world, I think there's a separation of responsibility between "function that defines how data is processed" and "thing that actually applies function to data." By analogy to an MVC framework, a separate model and controller.

Continuing that analogy, the choice to take a class gives some of the model responsibility to the controller itself, which for me in my own experience encourages leaky abstractions between the two and if the slippery slope is followed long enough, eventually both model and controller needing access to a broad array of global variables.

So coming to this file with my own expectations and biases, I expected the first argument to the function instead to be:
```python
   batch_processor: TorchBatchProcessor,
```

I talked to @azhou-determined looking for background for how the decision for the class argument came to be, but the code has evolved a bit since she last looked at it, and I'm just not sure what the right choice is, now.

Trade-offs from that conversation:
* pro-class: the context manager for the core_context is in `torch_batch_process`, meaning the caller has less work to do
* pro-class: the caller also doesn't have to write code to instantiate the object
* pro-object: more consistent with both TrainerAPI and CoreAPI
* pro-object: encourages separation of concerns between object creation and object application
* pro-class: function is already implemented this way
* pro-object: I haven't read through the tests yet, but I suspect testing is easier with this choice

I just don't have the necessary understanding yet of this file or system to say which choice is right, but I sure _want_ the object interface to be the better choice.

I don't know. What do you think, here, @szewaiyuen6?

Reply:
Thank you for the comment, below are my thoughts:
- I do not see the one way being obviously superior as there are trade-offs both ways. In fact both ways are used by other libraries (Ray vs PyTorch Lightning)
- This is an experimental API that we expect to get feedback from users on

I would be inclined keep the API as is and see what the users think.

Reply:
Discussed offline, keeping the API as is for now.